Add SPI to save web page resources to disk

szewai opened the pull request:
#### f730df710d6a5e003cc8bdcabf4ea9f6a607d699
<pre>
Add SPI to save web page resources to disk
<a href="https://bugs.webkit.org/show_bug.cgi?id=262128">https://bugs.webkit.org/show_bug.cgi?id=262128</a>
rdar://116069722

Reviewed by Ryosuke Niwa and David Kilzer.

WKWebView has API (createWebArchiveDataWithCompletionHandler) to store web page in webarchive format, which can only be
understood by WebKit. This patch introduces a new SPI to allow storing web page and its subresources in file format that
matches their MIME type, so the saved page can be loaded in other non-WebKit browsers.

Similar to the previous API, the new SPI makes UI process request webarchive data of a web page from web process.
However, instead of returning the data to API client, UI process will decode the data and store resources to given
directory. Also, when creating web archive of web page, web process replaces urls in the web page with file names, so
the saved page will point to saved subresource files instead of network resources.

API test: WebArchive.SaveResourcesBasic

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::MarkupAccumulator):
(WebCore::MarkupAccumulator::resolveURLIfNeeded const):
* Source/WebCore/editing/MarkupAccumulator.h:
(WebCore::MarkupAccumulator::MarkupAccumulator):
* Source/WebCore/editing/markup.cpp:
(WebCore::serializeFragment):
* Source/WebCore/editing/markup.h:
(WebCore::serializeFragment):
* Source/WebCore/loader/archive/Archive.cpp:
(WebCore::Archive::saveResourcesToDisk):
* Source/WebCore/loader/archive/Archive.h:
* Source/WebCore/loader/archive/ArchiveError.h: Added.
(WebCore::errorDescription):
* Source/WebCore/loader/archive/ArchiveResource.cpp:
(WebCore::ArchiveResource::ArchiveResource):
(WebCore::ArchiveResource::create):
(WebCore::ArchiveResource::saveToDisk):
* Source/WebCore/loader/archive/ArchiveResource.h:
(WebCore::ArchiveResource::create):
(WebCore::ArchiveResource::fileName const):
(WebCore::ArchiveResource::setFileName):
(WebCore::ArchiveResource::ArchiveResource):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::createPropertyListRepresentation):
(WebCore::LegacyWebArchive::createResource):
(WebCore::LegacyWebArchive::create):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _saveResources:suggestedFileName:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::sharedFileQueue):
(WebKit::WebPageProxy::saveResources):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::webArchiveData):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
(WebKit::WebFrame::webArchiveData):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getWebArchiveOfFrameWithFileName):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: <a href="https://commits.webkit.org/268614@main">https://commits.webkit.org/268614@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7245949337304c62002df3cf0c9b4d2708c41eb2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20147 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20582 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21205 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22047 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18802 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20378 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23821 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20744 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20264 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20367 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20270 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17503 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22897 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17439 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18303 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24560 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18516 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18479 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22546 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19075 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/16192 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18276 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4837 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22617 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18900 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/loader/archive/Archive.cpp:
```diff
@@ -50,4 +50,27 @@ void Archive::clearAllSubframeArchives(HashSet<Archive*>& clearedArchives)
+    for (auto subframeArchive : m_subframeArchives) {
+        if (!subframeArchive->saveResourcesToDisk(directory))
+            return false;
+    }
```
When we return `false`, I'm curious how we clean up partially written output.

I didn't see any code to clean up partially written output, so I think we need to do something here--unless it's always written to a temp directory that can be cleaned up by the system.

ddkilzer commented on Source/WebCore/loader/archive/ArchiveResource.h:
```diff
@@ -36,22 +36,25 @@ class ArchiveResource : public SubstituteResource {
     WEBCORE_EXPORT static RefPtr<ArchiveResource> create(RefPtr<FragmentedSharedBuffer>&&, const URL&,
-        const String& mimeType, const String& textEncoding, const String& frameName,
-        const ResourceResponse& = ResourceResponse());
+    const String& mimeType, const String& textEncoding, const String& frameName, const ResourceResponse& = ResourceResponse(), const String& fileName = { });
```
Nit: The arguments should be all on one line, or they should be indented.  IMO, it's hard to read when arguments continue on a second line, but also line up with new functions.

If indenting triggers a `check-webkit-style` warning, we should fix that instead (but we do not have to fix that with this PR).

ddkilzer commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -225,7 +231,13 @@ RefPtr<ArchiveResource> LegacyWebArchive::createResource(CFDictionaryRef diction
 
-    return ArchiveResource::create(SharedBuffer::create(resourceData), URL { url }, mimeType, textEncoding, frameName, response);
+    auto fileName = static_cast<CFStringRef>(CFDictionaryGetValue(dictionary, LegacyWebArchiveResourceFileNameKey));
+    if (fileName && CFGetTypeID(fileName) != CFStringGetTypeID()) {
```
We should use `dynamic_cf_cast<CFStringRef>()` here instead of `static_cast<CFStringRef>()` in case we get an unexpected object type, then we can get rid of the manual `CFGetTypeID()` check!

```suggestion
    auto fileName = dynamic_cf_cast <CFStringRef>(CFDictionaryGetValue(dictionary, LegacyWebArchiveResourceFileNameKey));
    if (!fileName) {
```

ddkilzer commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3320,6 +3320,14 @@ - (void)_setObservedRenderingProgressEvents:(_WKRenderingProgressEvents)observed
 
+- (void)_saveResources:(NSURL *)directory suggestedFileName:(NSString *)name completionHandler:(void (^)(BOOL))completionHandler
+{
+    THROW_IF_SUSPENDED;
```
I think we need to call `completionHandler(false)` here if we're going to throw--unless throwing results in a crash, then we don't care.  :)

ddkilzer commented on Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
```diff
@@ -237,6 +237,7 @@ GenerateSyntheticEditingCommand(enum:uint8_t WebKit::SyntheticEditingCommandType
     GetSelectionAsWebArchiveData() -> (std::optional<IPC::SharedBufferReference> data)
     GetSourceForFrame(WebCore::FrameIdentifier frameID) -> (String string)
     GetWebArchiveOfFrame(WebCore::FrameIdentifier frameID) -> (std::optional<IPC::SharedBufferReference> dataReference)
+    GetWebArchiveOfFrameWithFileName(WebCore::FrameIdentifier frameID, String fileName) -> (std::optional<IPC::SharedBufferReference> dataReference)
```
So this new message will send a whole webarchive over CoreIPC.

@beidson and/or @achristensen07 should definitely weigh in on this part.

I wonder if we can't do all the writing in the WebContent process (not the UI process--if I understand the PR correctly).

ddkilzer replied:
Sihui reminded me that we're already sending webarchive plists over CoreIPC for...saving WebArchives, so this really isn't new (just a new endpoint).

ddkilzer reviewed (changes requested):
I think the following changes need to be made:
- Add a mechanism to clean up partially written output.
- Switch to use `dynamic_cf_cast<>()` in the one comment.

I also think Brady or Alex need to review this to make sure that adding the new CoreIPC message is okay.

ddkilzer reviewed (approved):
r=me with my previous concerns addressed.

Sihui reminded me that we're already sending webarchive plist over CoreIPC for the webarchive format itself, so this just adds a new endpoint.  (I guess it would be nice if we could reuse the same endpoint for both and pass an enum value for the requested format, but maybe that's more work than it's worth.)

rhoule-apple commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -308,6 +308,7 @@ for this property.
 - (void)_killWebContentProcess;
 - (void)_killWebContentProcessAndResetState;
 
+- (void)_saveResources:(NSURL *)directory suggestedFileName:(NSString *)name completionHandler:(void (^)(BOOL))completionHandler;
```
Two comments.

1. We should avoid returning boolean for indication of success or failure. If it fails, that would be great if we could get the exact error. (ie. Permission error?, Disk full?, etc...)
2. About `suggestedFileName`. What if the suggested file name can't be used? Does it returns a failure, or it picks another filename. If the filename is changed, the caller should be notified.

About the sub resources, how do I know where are located? Is it a folder with with the very same name (minus the extension) ?

szewai replied:
> We should avoid returning boolean for indication of success or failure. If it fails, that would be great if we could get the exact error. (ie. Permission error?, Disk full?, etc...)

Sure, I could make the SPI return an NSError*.

> About suggestedFileName. What if the suggested file name can't be used? Does it returns a failure, or it picks another filename. If the filename is changed, the caller should be notified.

What do you mean by "filename can't be used"? Do you mean the name is malformed or file with same name already exists?

> About the sub resources, how do I know where are located? Is it a folder with the very same name (minus the extension) ?
By default, the subresources are stored in suggestedFileName_files, e.g. if you pass "apple", then in the directory, you should have file apple.html and directory apple_files that holds subresources.

rhoule-apple replied:
> What do you mean by "filename can't be used"? Do you mean the name is malformed or file with same name already exists?

If I pass "apple.html" for the suggestedFileName. What happens if the file already exists? Does it become "apple (1).html"? Does it fail? What if the "apple_files" exists?

szewai replied:
Hmm with current implementation we will fill the extension, so you should pass "apple", and we will save it as "apple.html". The operation can succeed and the existing files will be overwritten.

szewai committed:
Add SPI to save web page resources to disk
https://bugs.webkit.org/show_bug.cgi?id=262128
rdar://116069722

Reviewed by Ryosuke Niwa and David Kilzer.

WKWebView has API (createWebArchiveDataWithCompletionHandler) to store web page in webarchive format, which can only be
understood by WebKit. This patch introduces a new SPI to allow storing web page and its subresources in file format that
matches their MIME type, so the saved page can be loaded in other non-WebKit browsers.

Similar to the previous API, the new SPI makes UI process request webarchive data of a web page from web process.
However, instead of returning the data to API client, UI process will decode the data and store resources to given
directory. Also, when creating web archive of web page, web process replaces urls in the web page with file names, so
the saved page will point to saved subresource files instead of network resources.

API test: WebArchive.SaveResourcesBasic

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::MarkupAccumulator):
(WebCore::MarkupAccumulator::resolveURLIfNeeded const):
* Source/WebCore/editing/MarkupAccumulator.h:
(WebCore::MarkupAccumulator::MarkupAccumulator):
* Source/WebCore/editing/markup.cpp:
(WebCore::serializeFragment):
* Source/WebCore/editing/markup.h:
(WebCore::serializeFragment):
* Source/WebCore/loader/archive/Archive.cpp:
(WebCore::Archive::saveResourcesToDisk):
* Source/WebCore/loader/archive/Archive.h:
* Source/WebCore/loader/archive/ArchiveError.h: Added.
(WebCore::errorDescription):
* Source/WebCore/loader/archive/ArchiveResource.cpp:
(WebCore::ArchiveResource::ArchiveResource):
(WebCore::ArchiveResource::create):
(WebCore::ArchiveResource::saveToDisk):
* Source/WebCore/loader/archive/ArchiveResource.h:
(WebCore::ArchiveResource::create):
(WebCore::ArchiveResource::fileName const):
(WebCore::ArchiveResource::setFileName):
(WebCore::ArchiveResource::ArchiveResource):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::createPropertyListRepresentation):
(WebCore::LegacyWebArchive::createResource):
(WebCore::LegacyWebArchive::create):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _saveResources:suggestedFileName:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::sharedFileQueue):
(WebKit::WebPageProxy::saveResources):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::webArchiveData):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
(WebKit::WebFrame::webArchiveData):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getWebArchiveOfFrameWithFileName):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: https://commits.webkit.org/268614@main
